hotfix: return error instead of log at `FromMapAndOption`

chansuke opened the pull request:
Make [FromMapAndOption](https://github.com/kubernetes-sigs/kustomize/blob/master/api/resource/factory.go#L64-L65) to return error.

DiptoChakrabarty commented on api/internal/plugins/utils/utils_test.go:
```diff
@@ -89,7 +89,8 @@ func TestUpdateResourceOptions(t *testing.T) {
 		err := in.Append(makeConfigMap(rf, name, c.behavior, c.hashValue))
 		require.NoError(t, err)
-		err = expected.Append(makeConfigMapOptions(rf, name, c.behavior, !c.needsHash))
+		config := makeConfigMap(rf, name, c.behavior, c.hashValue)
```
it seems you have replaced the `makeConfigMapOptions` with `makeConfigMap` function which is causing a linter error.

Additionally this is eliminating the test for  `makeConfigMapOptions` function.

Could you fix this.

chansuke replied:
@DiptoChakrabarty Thank you so much for your review.I will fix this.

DiptoChakrabarty commented:
also seems like the tests are failing can you go ahead and fix them

koba1t commented on api/internal/plugins/utils/utils_test.go:
```diff
@@ -89,7 +89,11 @@ func TestUpdateResourceOptions(t *testing.T) {
-		err = expected.Append(makeConfigMapOptions(rf, name, c.behavior, !c.needsHash))
+		config, err := makeConfigMapOptions(rf, name, c.behavior, !c.needsHash)
+		if err != nil {
+			t.Errorf("unexpected error: %v", err)
```
I think the `unexpected error` message is confusing.
Could you add any meaningful message?

koba1t commented on api/resmap/reswrangler_test.go:
```diff
@@ -836,11 +837,15 @@ func makeMap1() ResMap {
-		}))
+		})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
```
same here

koba1t commented on api/resmap/reswrangler_test.go:
```diff
@@ -854,14 +859,19 @@ func makeMap2(b types.GenerationBehavior) ResMap {
-		}))
+		})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
```
same too

koba1t commented on api/resmap/reswrangler_test.go:
```diff
@@ -874,24 +884,28 @@ func TestAbsorbAll(t *testing.T) {
-	assert.NoError(t, w.AbsorbAll(makeMap2(types.BehaviorMerge)))
+		})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
```
same too

koba1t commented on api/resmap/reswrangler_test.go:
```diff
@@ -874,24 +884,28 @@ func TestAbsorbAll(t *testing.T) {
-	err := w.AbsorbAll(w2)
+	w = makeMap1(t)
+	w2 = makeMap2(t, types.BehaviorUnspecified)
+	err = w.AbsorbAll(w2)
 	assert.Error(t, err)
```
If you can, Could you summarize it in one line?

koba1t commented on api/resource/factory.go:
```diff
@@ -42,7 +42,11 @@ func (rf *Factory) Hasher() ifc.KustHasher {
-	return rf.FromMapAndOption(m, nil)
+	res, err := rf.FromMapAndOption(m, nil)
+	if err != nil {
+		log.Fatal("Option must not be null")
```
It looks like this line abandoned a non-nil error.
Could you make a change to print the substance of the returned error?

koba1t commented on api/resource/factory.go:
```diff
@@ -52,19 +56,21 @@ func (rf *Factory) FromMapWithName(n string, m map[string]interface{}) *Resource
-	r := rf.FromMapAndOption(m, nil)
+	r, err := rf.FromMapAndOption(m, nil)
+	if err != nil {
+		log.Fatal("Option must not be null")
```
same here

koba1t reviewed (changes requested):
I think we should avoid the `fatal` function.
But we are currently using `log.Fatal()`, and we can fix it with the next PR.
https://github.com/kubernetes-sigs/kustomize/pull/5381/files#diff-9041e805f414eb549e8ebcf4f2b82c1f40403e3ebe86a1f67f9c76718bd91797L65

I think this PR is moving forward to the first step and is useful for kustomize.

@chansuke 
Thanks for your contribution!
I added any comments, Could you check it?

chansuke commented:
@koba1t I appreciate your feedback and review! I will go fix them.

chansuke commented:
@koba1t I've addressed your comments and made the necessary changes. Please take a look :pray:
And I'd like to work on this in the next PR if it's possible :bow:

> I think we should avoid the fatal function.
But we are currently using log.Fatal(), and we can fix it with the next PR.

natasha41575 commented:
/ok-to-test

koba1t commented:
/label tide/merge-method-squash

koba1t commented:
@chansuke 

Thanks for the fix!
Looks good for me!

/lgtm

@natasha41575 
Would you like to review this PR, or do you have any opinions?
In the meantime, I look forward to your reply.

natasha41575 commented:
@koba1t Thanks! I'll take a look on friday

natasha41575 commented on api/resource/factory.go:
```diff
@@ -42,7 +42,11 @@ func (rf *Factory) Hasher() ifc.KustHasher {
-	return rf.FromMapAndOption(m, nil)
+	res, err := rf.FromMapAndOption(m, nil)
+	if err != nil {
+		log.Fatalf("failed to create resource from map: %v", err)
```
I think the idea of the TODO was to return the error all the way to the top, e.g. completely avoid the use of `log.Fatalf` for this. That would mean `FromMap` and `FromMapWithNamespaceAndName` should the error, instead of doing `log.Fatalf` here, and likewise all the way up.

Looks like you and @koba1t already had a discussion about potentially doing that in a follow-up PR. I think that's fine to do in a followup, but in that case could we keep the TODO here? 

```	
if err != nil {
		// TODO: return err instead of log.
		log.Fatalf("failed to create resource from map: %v", err)
	}
```

chansuke replied:
@natasha41575 Thank you for the review. Yes, I will put the TODO comment again

natasha41575 commented on api/resource/factory.go:
```diff
@@ -52,19 +56,21 @@ func (rf *Factory) FromMapWithName(n string, m map[string]interface{}) *Resource
-	r := rf.FromMapAndOption(m, nil)
+	r, err := rf.FromMapAndOption(m, nil)
+	if err != nil {
+		log.Fatalf("failed to create resource from map: %v", err)
```
same here

chansuke commented:
@natasha41575 Updated the code. Please take a look.

natasha41575 reviewed (commented):
/lgtm
/approve
